tests/rptest/tests/datalake/datalake_e2e_test.py

```diff
@@ -325,6 +327,78 @@ def assert_location_prefix(rows, prefix: str):
                 f"select path from {table_name}.manifests")
             assert_location_prefix(manifests, table.location())
 
+    @cluster(num_nodes=3)
+    @matrix(cloud_storage_type=supported_storage_types(),
+            catalog_type=supported_catalog_types(),
+            custom_partition_spec=[None, "(timestamp_us)", "(number)"])
+    def test_iceberg_partition_key_file_location(self, cloud_storage_type,
+                                                 catalog_type,
+                                                 custom_partition_spec: str):
+        """
+        Test that the data file location includes the partition key
+        """
+        count = 100
+        with DatalakeServices(self.test_ctx,
+                              redpanda=self.redpanda,
+                              catalog_type=catalog_type,
+                              include_query_engines=[QueryEngineType.SPARK
+                                                     ]) as dl:
+            config = {}
+            if custom_partition_spec:
+                config[
+                    "redpanda.iceberg.partition.spec"] = custom_partition_spec
+
+            dl.create_iceberg_enabled_topic(
+                self.topic_name,
+                partitions=2,
+                iceberg_mode="value_schema_id_prefix",
+                config=config)
+
+            schema = avro.loads(avro_schema_str)
+            producer = AvroProducer(
+                {
+                    'bootstrap.servers': self.redpanda.brokers(),
+                    'schema.registry.url':
+                    self.redpanda.schema_reg().split(",")[0]
+                },
+                default_value_schema=schema)
+            for _ in range(count):
+                t = time.time()
+                record = {"number": int(t), "timestamp_us": int(t * 1000000)}
+                producer.produce(topic=self.topic_name, value=record)
+
+            producer.flush()
+            dl.wait_for_translation(self.topic_name, msg_count=count)
+
+            spark = dl.spark()
+            table_name = f"redpanda.{self.topic_name}"
+            uri_pattern = re.compile(
+                r"(?P<scheme>.*?)://(?P<bucket>.*?)/(?P<key>.*)")
+
+            def validate_data_file_path(file_url):
+                m = uri_pattern.match(file_url)
+                assert m, f"Expected file url to match URI pattern: {file_url}"
+                assert m[
+                    'bucket'] == self.si_settings.cloud_storage_bucket, f"Expected bucket {m['bucket']} to be {self.si_settings.cloud_storage_bucket}"
+
+                path_parts = m['key'].split("/")
+                partition_key = path_parts[3]
+                dt = datetime.datetime.now()
```

Comment:
Shouldn't this be something like, the `datetime.datetime.now()` from right after calling our first produce? Otherwise it seems like this could be flaky when the year changes :)

Reply:
yeah, this is not perfect around the new years eve